Migrating cluster membership from ember to vue (clusterRoleTemplateBindings)

codyrancher opened the pull request:
![image](https://user-images.githubusercontent.com/55104481/118563816-1d1bd100-b724-11eb-9d0d-c89bc91b9907.png)
![image](https://user-images.githubusercontent.com/55104481/118563830-22791b80-b724-11eb-9b9c-37ab7f4bf11b.png)

rancher/dashboard#2501

richard-cox commented on config/product/explorer.js:
```diff
@@ -209,6 +215,17 @@ export function init(store) {
   });
 
+  virtualType({
+    label:       'Cluster Members',
```
Localisation

codyrancher replied:
Switched this label and the label of the other virtual types.

richard-cox reviewed (changes requested):
- In the members table do we need to group these by namespace? I think there'll only ever be one (is it some kind of reference to the cluster?)
- I've got some duff memberships which were based on users and groups from providers that are no longer available. When viewing the Cluster Membership table these cause an error
  ```
  client.js?06a0:97 TypeError: Cannot read property 'username' of undefined
    at Proxy.userAvatar (management.cattle.io.clusterroletemplatebinding.js?b8c7:27)
  ```

  In ember
  ![image](https://user-images.githubusercontent.com/18697775/118807453-3cd80400-b8a0-11eb-9c2d-25cadfa3eeb2.png)
 (the users vs principal thing might help here, see inline comments)

- This is a real PITA request, but the verbage of 'Create' for Cluster Members may get confusing compared to 'Add'. I went through this fun with users&auth/groups. In the end there needs to be a custom `list/members` where you can define a custom button for 'Add' and then a custom edit location/page to allow the different verbage in the title (though this might be achievable by some kind of mock 'mode'). Might be worth estimating effort then checking with Neil.

codyrancher commented on edit/management.cattle.io.clusterroletemplatebinding.vue:
```diff
@@ -0,0 +1,174 @@
+    @finish="saveOverride"
+    @cancel="done"
+  >
+    <SelectPrincipal class="mb-20" :mode="mode" :retain-selection="true" @add="onAdd" />
```
Yeah, I had mentioned it. When looking at it I realized that we'd have to resize the component depending on whether or not something was or wasn't selected and I really dislike elements that I interact changing the layout so I decided to not change it.

codyrancher commented on models/management.cattle.io.clusterroletemplatebinding.js:
```diff
@@ -1,6 +1,37 @@
+  },
+
+  user() {
+    return this.$rootGetters['management/byId'](MANAGEMENT.USER, this.userName);
```
Switched to Principal

richard-cox replied:
There's some leftovers here from the previous approach
- users could probably be removed
- principal should be updated (throws error when adding a group to a cluster). This think could just be `userPrincipalId` if just the id is needed
- nameDisplay will change to the principal's nameDisplay

codyrancher commented on models/management.cattle.io.user.js:
```diff
@@ -17,6 +19,29 @@ export default {
+      .filter(p => p);
+  },
+
+  avatarSrc() {
```
Removed

codyrancher commented on models/namespace.js:
```diff
@@ -100,7 +100,7 @@ export default {
 
   projectNameSort() {
-    return this.project?.nameSort || '';
+    return this.project?.nameSort || '}';
```
Switched back.

codyrancher commented on edit/management.cattle.io.clusterroletemplatebinding.vue:
```diff
@@ -0,0 +1,174 @@
+    </h3>
+    <label class="mt-0 mb-0">{{ t('members.clusterPermissions.description') }}</label>
+    <div class="mb-10 mt-20">
+      <label class="radio">
```
Finished

codyrancher commented on edit/management.cattle.io.clusterroletemplatebinding.vue:
```diff
@@ -0,0 +1,174 @@
+      <label class="radio">
+        <input v-model="permissionGroup" :disabled="isView" type="radio" value="owner" />
+        {{ t('members.clusterPermissions.owner.label') }}
+        <div class="text-small text-muted">{{ t('members.clusterPermissions.owner.description') }}</div>
```
I switched to radio group and added description to the component.

codyrancher commented on edit/management.cattle.io.clusterroletemplatebinding.vue:
```diff
@@ -0,0 +1,174 @@
+    </div>
+    <div v-if="permissionGroup === 'custom'" class="custom-permissions">
+      <label v-for="permission in customPermissions" :key="permission.key" class="mb-5">
+        <input v-model="permission.value" :disabled="isView" type="checkbox" />
```
Switched to checkbox

codyrancher commented:
> 
> 
>     * In the members table do we need to group these by namespace? I think there'll only ever be one (is it some kind of reference to the cluster?)
Removed grouping

>     * I've got some duff memberships which were based on users and groups from providers that are no longer available. When viewing the Cluster Membership table these cause an error
>       ```
>       client.js?06a0:97 TypeError: Cannot read property 'username' of undefined
>         at Proxy.userAvatar (management.cattle.io.clusterroletemplatebinding.js?b8c7:27)
>       ```
I didn't verify against github but when I had things misconfigured I saw that the Principal formatter displayed something similar to ember.     
>         
>       In ember
>       ![image](https://user-images.githubusercontent.com/18697775/118807453-3cd80400-b8a0-11eb-9c2d-25cadfa3eeb2.png)
>       (the users vs principal thing might help here, see inline comments)
> 
>     * This is a real PITA request, but the verbage of 'Create' for Cluster Members may get confusing compared to 'Add'. I went through this fun with users&auth/groups. In the end there needs to be a custom `list/members` where you can define a custom button for 'Add' and then a custom edit location/page to allow the different verbage in the title (though this might be achievable by some kind of mock 'mode'). Might be worth estimating effort then checking with Neil.

Also made this change.


I also made it visually a little more consistent with your pages.
![image](https://user-images.githubusercontent.com/55104481/120051471-a27d6c00-bfd5-11eb-85e0-ed981a4fc4fd.png)

richard-cox commented on config/product/explorer.js:
```diff
@@ -237,18 +254,6 @@ export function init(store) {
-    group:            'cluster',
-    icon:             'globe',
-    namespaced:       false,
-    ifRancherCluster: false,
```
Think this should stay, it's the fallback if ifRancherCluster is false

codyrancher replied:
Yeah, I don't even remember removing this. Sorry.

richard-cox commented on config/product/explorer.js:
```diff
@@ -211,7 +217,18 @@ export function init(store) {
   virtualType({
-    label:          'Overview',
+    label:       store.getters['i18n/t']('members.clusterMembers'),
+    group:      'rbac',
```
I got in a pickle with this location by incorrectly assuming roles created in this RBAC section's `ClusterRoles` page should then apply to the roles shown when in `Cluster Members`. Minus membership all other sections are core kube `rbac.authorization.k8s.io` resource types. Feels like  `Cluster Membership` should move into the top level `Cluster` section alongside projects + nodes. Ideally everything role/membership related should be in RBAC but I don't think we can escape from the management cluster vs rbac cluster confusion

codyrancher replied:
I mean I put this here because as you said ideally everything role/membership related should be in RBAC. I'd prefer to organize this by concept and not by types. I'm going to leave this as is for now but if it does lead to more confusion I'll go ahead and change it.

codyrancher replied:
I'd rather keep this as is. As you mentioned ideally it's in the right spot. If it does add more confusion then I'll change it.

richard-cox commented:
Most comments resolved, anything new / remaining is still open

codyrancher commented on edit/management.cattle.io.clusterroletemplatebinding.vue:
```diff
@@ -0,0 +1,222 @@
+    },
+    options() {
+      const customRoles = this.customRoles.map(role => ({
+        label:       role.displayName,
```
done

codyrancher commented on edit/management.cattle.io.clusterroletemplatebinding.vue:
```diff
@@ -0,0 +1,222 @@
+    options() {
+      const customRoles = this.customRoles.map(role => ({
+        label:       role.displayName,
+        description: role.description,
```
I tried a few different variants. I decided to go with
![image](https://user-images.githubusercontent.com/55104481/120347900-887faa00-c2b1-11eb-9ece-00a3cfdb4549.png)

codyrancher commented:
It wouldn't let me reply inline

> - users could probably be removed

Users is still used to do filtering on the list view. Specifically to check if the user `isSystem`.

> - principal should be updated (throws error when adding a group to a cluster). This think could just be userPrincipalId if just the id is needed

I didn't notice this error. I added a group and it appears to create and fetch it just fine. It does look like I can remove principal though so I went ahead and did that.
![image](https://user-images.githubusercontent.com/55104481/120350693-0fce1d00-c2b4-11eb-8206-8beb3d868e8c.png)

> - nameDisplay will change to the principal's nameDisplay

I just removed the overload since we're not even using it.

richard-cox commented:
I tried reproducing the error on adding a group and couldn't. The only issue left is the binding's principal, it looks like it's correctly used in the `principal` table column for sorting. Search though looks broken, as the binding's `userPrincipalName` value will contain id's rather than names for a user (and possibly some groups). That value could also be `principal.loginName`. Fetching the principal via the binding's `userPrincipalName` should just work with a getter (as the Prinicpal formatter should fetch it).

codyrancher commented:
> 
> 
> I tried reproducing the error on adding a group and couldn't. The only issue left is the binding's principal, it looks like it's correctly used in the `principal` table column for sorting. Search though looks broken, as the binding's `userPrincipalName` value will contain id's rather than names for a user (and possibly some groups). That value could also be `principal.loginName`. Fetching the principal via the binding's `userPrincipalName` should just work with a getter (as the Prinicpal formatter should fetch it).

I just decided to make use of the sub-search and sub-fields props on sortable table to search for nameDisplay.

richard-cox commented:
> > I tried reproducing the error on adding a group and couldn't. The only issue left is the binding's principal, it looks like it's correctly used in the `principal` table column for sorting. Search though looks broken, as the binding's `userPrincipalName` value will contain id's rather than names for a user (and possibly some groups). That value could also be `principal.loginName`. Fetching the principal via the binding's `userPrincipalName` should just work with a getter (as the Prinicpal formatter should fetch it).
> 
> I just decided to make use of the sub-search and sub-fields props on sortable table to search for nameDisplay.

Ok, worked out some of the confusion. Looks like template binding's `user` is always populated even if the principal is a group. I didn't think this happened and that nameDisplay would be blank for groups.
